client/beefy/src/worker.rs

```diff
@@ -216,77 +262,211 @@ where
+
+				trace!(target: "beefy", "🥩 round concluded, waking up ticker");
+				self.ticker.wake();
+			}
+		}
+	}
+
+	/// Set best BEEFY block to `block_num`.
+	///
+	/// Also sends/updates the best BEEFY block hash to the RPC worker.
+	fn set_best_beefy_block(&mut self, block_num: NumberFor<B>) {
+		if Some(block_num) > self.best_beefy_block {
+			// Try to get block hash ourselves.
+			let block_hash = match self.client.hash(block_num) {
+				Ok(h) => h,
+				Err(e) => {
+					error!(target: "beefy", "🥩 Failed to get hash for block number {}: {}",
+						block_num, e);
+					None
+				},
+			};
+			// Update RPC worker with new best BEEFY block hash.
+			block_hash.map(|hash| {
+				self.beefy_best_block_sender
+					.notify(|| Ok::<_, ()>(hash))
+					.expect("forwards closure result; the closure always returns Ok; qed.")
+			});
+			// Set new best BEEFY block number.
+			self.best_beefy_block = Some(block_num);
+			metric_set!(self, beefy_best_block, block_num);
+		} else {
+			debug!(target: "beefy", "🥩 Can't set best beefy to older: {}", block_num);
+		}
+	}
+
+	/// Return first block number pertaining to same session as `header`.
+	fn session_start_for_header(&self, header: &B::Header) -> NumberFor<B> {
+		// Current header is also the session start.
+		if let Some(_) = find_authorities_change::<B>(header) {
+			return *header.number()
+		}
+
+		// Walk up the chain looking for the session change digest.
+		let mut header = header.clone();
+		let mut parent_hash = *header.parent_hash();
+		while parent_hash != Default::default() {
+			header = self
+				.client
+				.expect_header(BlockId::Hash(parent_hash))
+				// TODO: is this proof correct?
+				.expect("header always available when following parent hash; qed.");
+			if let Some(_) = find_authorities_change::<B>(&header) {
+				// Found the first header of the session.
+				info!(target: "beefy", "🥩 session boundary is: {:?}.", *header.number());
+				return *header.number()
+			}
+			parent_hash = *header.parent_hash();
+		}
+
+		info!(target: "beefy", "🥩 no session boundary found, defaulting to block number 1.");
+		1u32.into()
+	}
+
+	/// Handle potential session changes by starting new voting round for mandatory blocks.
+	fn handle_session_change(&mut self) {
+		let header = &self.best_grandpa_block_header;
+		if let Some(active) = self.validator_set(header) {
+			// If validator set has changed compared to the one in our active round,
 			if self.rounds.is_none() ||
-				active.id() != self.rounds.as_ref().unwrap().validator_set_id() ||
-				(active.id() == GENESIS_AUTHORITY_SET_ID && self.best_beefy_block.is_none())
+				active.id() != self.rounds.as_ref().unwrap().validator_set_id()
 			{
-				debug!(target: "beefy", "🥩 New active validator set id: {:?}", active);
-				metric_set!(self, beefy_validator_set_id, active.id());
+				// Find new session_start and start new round for mandatory block.
+				let session_start = self.session_start_for_header(header);
```

Comment:
If we assume that the block that changes authorities is always finalized by GRANDPA then `session_start` will always be `best_grandpa_block_header` (since this is now done asynchronously unfortunately we can't assume that `best_grandpa_block_header` has not moved forward in the meantime).